test/ipfs-api/pubsub.spec.js

```diff
@@ -0,0 +1,119 @@
+/* eslint-env mocha */
+/* eslint max-nested-callbacks: ['error', 8] */
+/* globals apiClients */
+'use strict'
+
+const expect = require('chai').expect
+const isNode = require('detect-node')
+const FactoryClient = require('../factory/factory-client')
+
+const topicName = 'js-ipfs-api-tests'
+
+const publish = (ipfs, data, callback) => {
+  ipfs.pubsub.publish(topicName, data, (err, successful) => {
+    expect(err).to.not.exist
+    expect(successful).to.equal(true)
+    callback()
+  })
+}
+
+describe('.pubsub', () => {
+  if (!isNode) {
+    return
+  }
+
+  let ipfs
+  let fc
+
+  before(function (done) {
+    fc = new FactoryClient()
+    fc.spawnNode((err, node) => {
+      expect(err).to.not.exist
+      if (err) done(err)
+      ipfs = node
+      done()
+    })
+  })
+
+  after((done) => {
+    fc.dismantle(done)
+  })
+
+  describe('.publish', () => {
+    it('message from string', (done) => {
+      publish(ipfs, 'hello friend', done)
+    })
+    it('message from buffer', (done) => {
+      publish(ipfs, new Buffer('hello friend'), done)
+    })
+  })
+
+  describe('.subscribe', () => {
+    it('one topic', (done) => {
+      ipfs.pubsub.subscribe(topicName, (err, subscription) => {
+        expect(err).to.not.exist
+        subscription.on('data', (d) => {
+          expect(d.data).to.equal('hi')
+          subscription.cancel()
+        })
+        subscription.on('end', () => {
+          done()
+        })
+      })
+      setTimeout(publish.bind(null, ipfs, 'hi', () => {}), 0)
+    })
+    it('fails when already subscribed', (done) => {
+      ipfs.pubsub.subscribe(topicName, (firstErr, firstSub) => {
+        expect(firstErr).to.not.exist
+        ipfs.pubsub.subscribe(topicName, (secondErr, secondSub) => {
+          expect(secondErr).to.be.an('error')
+          expect(secondErr.toString()).to.equal('Error: Already subscribed to ' + topicName)
```

Comment:
![selection_023](https://cloud.githubusercontent.com/assets/459764/18516929/06409d28-7a9b-11e6-83d0-e16337e24991.png)
Behavior copied from the go-ipfs implementation, not sure if needed or not. We could reuse the stream and abstract away this from js-ipfs-api, if you want?